### ISSUE TITLE
CDIInjectionEnricher does not work for glue classes

### ISSUE THREAD

**Issue by lsuski** [cukespace/cukespace#81]
Hi,
I'm using cukespace without drone extension beacuse I don't need UI testing but the problem is that dependencies are not injected in steps class declared in @Glue annotation.
CDIInjectionEnricher is invoked but no context is active so BeanManager cannot be obtained. I think that glue instantation should be perfomed in some context (probably ApplicationContext or all contexts).

**Comment by rmannibucau** [comment]
Hi

I think it depends your adapter, enrichment is typically done with glue loading (from cucumber) which is when the Runtime (cucumber) is created which happens in a @Test method so the context should be all setup.

**Comment by lsuski** [comment]
What do you mean by 'adapter'?
I have test class 
```java
@RunWith(CukeSpace.class)
@Features({
        "classpath:my_feature.feature"})
@Glues(CommonSteps.class)
public class TestSteps
{

       @Inject
	AppDefStore appDefStore;


    @Deployment
    public static WebArchive createArchiveAndDeploy()
    {
        return ShrinkWrap.create(WebArchive.class)
                .addClasses(
                        AppDefStore.class,         
                        TestSteps.class,
		                )
                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml");
    }
}
```
and 
```java
public class CommonSteps
 {
	@Inject
	AppDefStore appDefStore;


	@Given("^some common step$")
	public void someCommonStep() throws Throwable
	{
		appDefStore.registerAppDefs();
	}


 }
```

I don't have any @Test method because it is cucumber. Dependency AppDefStore is injected in TestSteps but not in CommonSteps

**Comment by rmannibucau** [comment]
adapter = the container integration you use with CDI

the @Test method is simulated in JUnit runtime by cukespace but it behaves the same way.

you can debug in cucumber.runtime.arquillian.backend.ArquillianBackend#loadGlue to check what is happening (cucumber.runtime.arquillian.backend.ArquillianBackend#initInstances to be precise)

**Comment by lsuski** [comment]
I use GlassFish Embedded 3.1.
I've already debugged ArquillianBackend and realized that CucumberLifecycle.enrich(instance) invokes CDIInjectionEnricher which invokes beanManagerInst.get() which goes to org.jboss.arquillian.core.impl.ManagerImpl.resolve(type). ManagerImpl have some contexts but none of them is active at this time.

**Comment by rmannibucau** [comment]
So the context is started with the Manager(Impl) by arquillian so this is not the issue. The only one i can think about is glassfish embedded not setting the Context (InitialContext, nothing linked to arquillian scopes) in arquillian so org.jboss.arquillian.testenricher.cdi.container.BeanManagerProducer#findBeanManager is never fired.

**Comment by lsuski** [comment]
Problem is with threads. ApplicationContextImpl used indirectly by CDIInjectionEnricher used by CucumberLifecycle is activated in main thread, but instantiation of glue takes place in some http-listener thread

**Comment by lsuski** [comment]
AbstractContext has ```ThreadLocal<Stack<StoreHolder<T>>>activeStore``` field which returns empty stack for http-listener thread and non-empty for main thread.
Do you suggest any workaround for this?

**Comment by rmannibucau** [comment]
The only workaround I can think about is a custom LoadableExtension using a static variable to store the value and propagating/re-setting it from each request (with a filter) but this sounds like a bug in glassfish adapter (or an incompatibility with last arquillian releases).

**Comment by lsuski** [comment]
I'm not sure if this is GlassFish adapter problem. 
CucumberLifecycle.loadCucumberAnnotationsAndEnrichers method is called twice. First time is in main thread and second one is in http-listener thread but then TEST_ENRICHERS.isEmpty() is false. CDIInjectionEnricher should be loaded second time with proper ApplicationContextImpl.

**Comment by rmannibucau** [comment]
Test enricher loading happens on client and server sides - cause they can be different - but has is not linked to any scope, it is just to load instances. Also it works fine on other adapters so likely a glassfish issue.

**Comment by lsuski** [comment]
So, maybe I don't understand how it should work. In which thread should be executed method `ArquillianCucumber.performInternalCucumberOperations` ? Because it is executed in some glassfish worker thread because internally it is HTTP POST request executed by `ServletMethodExecutor`. Event `org.jboss.arquillian.test.spi.event.suite.Before` is called twice. First time is invoked by junit from method `runChild`. Second one is invoked by `ServletTestRunner` (of course indirectly). I can't test it on different adapters so I don't know for sure it works on them. Could you tell me where is the problem in this process and why it is related with GlassFish? As tests are executed as HTTP request it is quite normal that they are executed in worker threads. According to AbstractContext implementation activness of context depends on thread and injection depends on context being active. In my opinion there shouldn't be two `Before` events. Could you clarify this?

**Comment by rmannibucau** [comment]
performInternal* is executed in a "test thread". For glassfish - using servlet protocol of arquillian - it is a worker thread. That said it doesn't prevent the arquillian setup to have to work, said otherwise the bean manager, initial context should be available for all worker threads while the application is deployed. Before is per test method, I wonder if your 2 counting is not client+server (since it is embedded you see both), can it be that?

**Comment by lsuski** [comment]
If I correctly understand client that first Before is initialized by JUnitStarter which goes to `Statement` returned by `Arquillian,withBefores`. Second Before is initialized by `org.jboss.arquillian.protocol.servlet.runner.ServletTestRunner` which goes to `org.jboss.arquillian.junit.container.JUnitTestRunner`  and again to  `Statement` returned by `Arquillian,withBefores` so we have 2 Before events. Everything works for main test class because `ManagerImpl` and enrichers are instantiated in different way than for glue classes - actually they are instantiated second time in worker thread and have new contexts which are active in this thread. For glues there is fdifferent situation because we have `if (TEST_ENRICHERS.isEmpty())` check which prevents TestEnrichers being instantiated second time. In turn for glue class there is used old `ManagerImpl` with old contexts active for main/test thread and for main test class there is used new `ManagerImpl` with different contexts which are active in worker thread. I've debugged this and I'm sure about it. So the question is - should `TEST_ENRICHERS.isEmpty()` check exists or should there be 2 Before events?

**Comment by rmannibucau** [comment]
That's a good point, you can try to patch cukespace remove the TEST_ENRICHERS.isEmpty() condition to recreate enricher each time.

With remote adapters this is not needed and with stateless enrichers of embedded adapters it is neither (tomee one for instance).

If it works this is something we can manage in cukespace.

**Comment by lsuski** [comment]
I've fixed this by adding `RemoteLoadableExtension` which resets `CucumberLifecycle`

```java
public void fix(@Observes(precedence = 2) Before before)
	{
		CucumberLifecycle lifecycle = new CucumberLifecycle();
		lifecycle.reset(null);
		lifecycle.init(null);
	}
```

**Comment by lsuski** [comment]
But there is another problem with lifespan of managed object. Generally in Cucumber there is a way to manage injected objects by defining if they can live longer than scenario (using custom ObjectFactory, e.g. cucumber-guice). This approach is very handy because you don't need to manually reset any state created by scenario. Here in cukespace `ArquillianBackend` has empty `build/disposeWorld` methods which should be used to clear any scenario state

**Comment by lsuski** [comment]
I think that this is generally problem of Arquillian and  ArquillianCucumber that tests are not fully isolated which is different than how JUnit works and what people are familiar with - I mean that each test nonstatic state should be reset. Here we have CDI which decides what is a lifespan of an object and e.g. `@RequestScoped` object lives across all tests which is really not intuitive for programmer and is caused by Arquillian internals.

**Comment by rmannibucau** [comment]
yes and no, it fully respects JUnit but then arquillian depends the protocol and adapter which is the not intuitive part.

Do you want to propose a PR removing the caching if enrichers and loading them each time?

**Comment by lsuski** [comment]
If it is good solution I can make a PR - if this caching was only for optimisation than it shouldn't break anything else?

**Comment by rmannibucau** [comment]
Yes, was just for performances. A ThreadLocal or InheritedThreaLocal would work too probably
